File: DevExpress-Demo/DevExpress.SalesDemo.Wpf/CS/DevExpress.SalesDemo.Model/Interfaces.cs
using System;
using System.Collections.Generic;

namespace DevExpress.SalesDemo.Model {
    public interface IDataProvider {
        SalesGroup GetTotalSalesByRange(DateTime start, DateTime end);
        IEnumerable<SalesGroup> GetSales(DateTime start, DateTime end, GroupingPeriod period);
        IEnumerable<SalesGroup> GetSalesByChannel(DateTime start, DateTime end, GroupingPeriod period);
        IEnumerable<SalesGroup> GetSalesByProduct(DateTime start, DateTime end, GroupingPeriod period);
        IEnumerable<SalesGroup> GetSalesByRegion(DateTime start, DateTime end, GroupingPeriod period);
        IEnumerable<SalesGroup> GetSalesBySector(DateTime start, DateTime end, GroupingPeriod period);
    }
    public static class DataProviderExtensions {
        public static SalesGroup GetTotalSalesByRange(this IDataProvider dataProvider, DateTimeRange range) {
            Verify(dataProvider);
            return dataProvider.GetTotalSalesByRange(range.Start, range.End);
        }
        public static IEnumerable<SalesGroup> GetSales(this IDataProvider dataProvider, DateTimeRange range, GroupingPeriod period) {
            Verify(dataProvider);
            return dataProvider.GetSales(range.Start, range.End, period);
        }
        public static IEnumerable<SalesGroup> GetSalesByChannel(this IDataProvider dataProvider, DateTimeRange range, GroupingPeriod period) {
            Verify(dataProvider);
            return dataProvider.GetSalesByChannel(range.Start, range.End, period);
        }
        public static IEnumerable<SalesGroup> GetSalesByProduct(this IDataProvider dataProvider, DateTimeRange range, GroupingPeriod period) {
            Verify(dataProvider);
            return dataProvider.GetSalesByProduct(range.Start, range.End, period);
        }
        public static IEnumerable<SalesGroup> GetSalesByRegion(this IDataProvider dataProvider, DateTimeRange range, GroupingPeriod period) {
            Verify(dataProvider);
            return dataProvider.GetSalesByRegion(range.Start, range.End, period);
        }
        public static IEnumerable<SalesGroup> GetSalesBySector(this IDataProvider dataProvider, DateTimeRange range, GroupingPeriod period) {
            Verify(dataProvider);
            return dataProvider.GetSalesBySector(range.Start, range.End, period);
        }

        static void Verify(IDataProvider dataProvider) {
            if(dataProvider == null)
                throw new ArgumentNullException();
        }
    }
    //
    public interface IDataGenerator {
        event ProgressEventHandler GenerationStart;
        event ProgressEventHandler GenerationProgress;
        event ProgressEventHandler GenerationComplete;
    }
    public delegate void ProgressEventHandler(
            object sender, ProgressEventArgs e
        );
    public class ProgressEventArgs {
        public ProgressEventArgs(int progress) {
            Progress = progress;
        }
        public int Progress { get; private set; }
    }
}

File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/VB/DevExpress.ProductsDemo.Wpf/Sales.Model/Interfaces.vb
Imports System
Imports System.Collections.Generic
Imports System.Runtime.CompilerServices

Namespace DevExpress.SalesDemo.Model

    Public Interface IDataProvider

        Function GetTotalSalesByRange(ByVal start As Date, ByVal [end] As Date) As SalesGroup

        Function GetSales(ByVal start As Date, ByVal [end] As Date, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)

        Function GetSalesByChannel(ByVal start As Date, ByVal [end] As Date, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)

        Function GetSalesByProduct(ByVal start As Date, ByVal [end] As Date, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)

        Function GetSalesByRegion(ByVal start As Date, ByVal [end] As Date, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)

        Function GetSalesBySector(ByVal start As Date, ByVal [end] As Date, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)

    End Interface

    Public Module DataProviderExtensions

        <Extension()>
        Public Function GetTotalSalesByRange(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange) As SalesGroup
            Verify(dataProvider)
            Return dataProvider.GetTotalSalesByRange(range.Start, range.End)
        End Function

        <Extension()>
        Public Function GetSales(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)
            Verify(dataProvider)
            Return dataProvider.GetSales(range.Start, range.End, period)
        End Function

        <Extension()>
        Public Function GetSalesByChannel(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)
            Verify(dataProvider)
            Return dataProvider.GetSalesByChannel(range.Start, range.End, period)
        End Function

        <Extension()>
        Public Function GetSalesByProduct(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)
            Verify(dataProvider)
            Return dataProvider.GetSalesByProduct(range.Start, range.End, period)
        End Function

        <Extension()>
        Public Function GetSalesByRegion(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)
            Verify(dataProvider)
            Return dataProvider.GetSalesByRegion(range.Start, range.End, period)
        End Function

        <Extension()>
        Public Function GetSalesBySector(ByVal dataProvider As IDataProvider, ByVal range As DateTimeRange, ByVal period As GroupingPeriod) As IEnumerable(Of SalesGroup)
            Verify(dataProvider)
            Return dataProvider.GetSalesBySector(range.Start, range.End, period)
        End Function

        Private Sub Verify(ByVal dataProvider As IDataProvider)
            If dataProvider Is Nothing Then Throw New ArgumentNullException()
        End Sub
    End Module

    Public Interface IDataGenerator

        Event GenerationStart As ProgressEventHandler

        Event GenerationProgress As ProgressEventHandler

        Event GenerationComplete As ProgressEventHandler

    End Interface

    Public Delegate Sub ProgressEventHandler(ByVal sender As Object, ByVal e As ProgressEventArgs)

    Public Class ProgressEventArgs

        Private _Progress As Integer

        Public Sub New(ByVal progress As Integer)
            Me.Progress = progress
        End Sub

        Public Property Progress As Integer
            Get
                Return _Progress
            End Get

            Private Set(ByVal value As Integer)
                _Progress = value
            End Set
        End Property
    End Class
End Namespace
